firedrake/slate/slac/optimise.py

```diff
@@ -0,0 +1,55 @@
+from gem.node import MemoizerArg
+from functools import singledispatch
+from itertools import repeat
+import firedrake.slate.slate as sl
+
+
+def optimise(expression):
+    return push_block(expression)
+
+
+def push_block(expression):
+    mapper = MemoizerArg(_push_block)
+    ret = mapper(expression, ())
+    return ret
+
+
+@singledispatch
+def _push_block(expr, self, indices):
+    raise AssertionError("Cannot handle terminal type: %s" % type(expr))
+
+
+@_push_block.register(sl.Transpose)
+def _push_block_transpose(expr, self, indices):
+    return sl.Transpose(*map(self, expr.children, repeat(indices[::-1]))) if indices else expr
+
+
+@_push_block.register(sl.Add)
+@_push_block.register(sl.Negative)
+def _push_block_add(expr, self, indices):
+    return type(expr)(*map(self, expr.children, repeat(indices))) if indices else expr
+
+
+@_push_block.register(sl.Factorization)
+@_push_block.register(sl.Inverse)
+@_push_block.register(sl.Solve)
+@_push_block.register(sl.Mul)
+def _push_block_stop(expr, self, indices):
+    return sl.Block(expr, indices) if indices else expr
+
+
+@_push_block.register(sl.Tensor)
+def _push_block_tensor(expr, self, indices):
+    return sl.Tensor(sl.Block(expr, indices).form) if indices else expr
+
+
+@_push_block.register(sl.AssembledVector)
+def _push_block_assembled_vector(expr, self, indices):
+    return sl.BlockAssembledVector(sl.Block(expr, indices).form, indices) if indices else expr
+
+
+@_push_block.register(sl.Block)
+def _push_block_block(expr, self, indices):
+    indices = expr._indices if not indices else tuple(big[slice(*small)] for small, big in zip(expr._indices, indices))
+    block, = map(self, expr.children, repeat(indices))
```

Comment:
Maybe just:
```
return self(expr.children[0], indices)
```
?

Reply:
Why is that better? Do you think `map` & `repeat` come with too much overhead if there is only one child?

Also I thought in general we prefer to destruct tuples with `child, = expr.children` rather than indexing. So maybe I should change this to the following instead?
```
@_push_block.register(sl.Block)
def _push_block_block(expr, self, indices):
    indices = expr._indices if not indices else tuple(big[slice(*small)] for small, big in zip(expr._indices, indices))
    child, = expr.children
    return self(child, indices)
```

Reply:
I think I was confused the first time I read things because I didn't parse the `map(..., repeat)` paradigm as `(self(child, indices) for child in self.children)`

Reply:
I don't care about the increased computational cost.

Reply:
Ah ok I will leave it as is then